---
title: "ggtree-PPR"
output: html_document
date: "18-03-2024"
---

## import R packages

```{r}
library(phytools)
library(ggplot2)
library(ggtree)
library(reshape2)
library(openxlsx)
```

## load the phylogenetic tree

```{r,include=FALSE}

# Set the working directory
path = "/Users/lfeng/Library/CloudStorage/OneDrive-NorwichBioScienceInstitutes/01.projects/01.PPRsiRNA/02.ath8acc/script/"
setwd(path)

# RING1: Read the phylogenetic tree
treefile = paste(path,"ath_475ppr.pep.fa.phy.contree",sep="")
tree <- read.tree(treefile)

p = ggtree(tree, aes(colour = tree$Color), color = "black", layout = "fan", size = 1.5) + 
      geom_tiplab2(offset = .08, size = 0, color = "grey", align = TRUE, linetype = NA) 
p = open_tree(p,8) 

# Define genes within the clade for highlighting
gene_list <- c("AT1G63330", "AT1G62590", "AT1G62910", "AT1G63150", "AT1G62914", 
               "AT1G63130", "AT1G62930", "AT1G63080", "AT1G63070", "AT1G63400", 
               "AT1G62670", "AT1G62720", "AT1G62680", "AT3G16710", "AT5G16640", 
               "AT1G06580", "AT1G64583", "AT1G12300", "AT1G12620", "AT1G12775", 
               "AT3G22470", "AT1G12700", "AT1G63230", "AT1G64100", "AT1G63630", 
               "AT3G17370", "AT4G26800", "AT1G64580", "AT5G41170")

hgene <- rep("gene", length(tree$tip.label))
hgene[tree$tip.label %in% gene_list] <- "highlight"
hgene.df <- data.frame(taxa = tree$tip.label, hgene = factor(hgene))

# Highlight genes within the clade
p <- p %<+% hgene.df + 
      geom_tippoint(aes(color = hgene), size = 0) +
      geom_tiplab2(aes(color = hgene), align = TRUE, size = 0, linetype = "dotted", linesize = 1.5) +
      scale_color_manual(values = c("gene" = "white", "highlight" = "white"), guide = 'none')

# RING2: Read chromosome data from Excel file
chr_data <- read.xlsx("TPM_21_PPR_all_accs.xlsx", sheet = 3, sep = "\t")

# Create the "Count" variable
chr_data$Count <- 1
chr_data$Count <- factor(chr_data$Count)

# Create the "Chromosome" variable as a factor
chr_data$Chromosome <- factor(chr_data$Chromosome)

# Add chromosome information
p = p + new_scale_fill() +
  geom_fruit(data = chr_data, 
             geom = geom_tile,
             mapping = aes(y = ID, x = Count, fill = Chromosome),
             color = "white", offset = 0.02, width = 0.2) +
  scale_fill_manual(values=c("1" = "#9E2214", "2" = "#4363d8", "3" = "#DA8932", "4" = "#390996", "5" = "#8EF6C4"),
                    guide = guide_legend(keywidth = 2, keyheight = 2, order = 4))

addline = data.frame(ID = chr_data$ID, Count="1")
addline$Count = factor(addline$Count)
p = p + new_scale_fill() +
  geom_fruit(data = addline, 
             geom = geom_tile,
             mapping = aes(y = ID, x = Count),
             color = "black", fill = "black", offset = 0.015, width = 0.02)

# RING3: Add miR target information
tpm_data <- read.xlsx("TPM_21_PPR_all_accs.xlsx", sheet = 2, sep = "\t")
tar_data <- read.delim2("target_region.info.txt", sep = "\t", header = TRUE)
tar_data <- tar_data[tar_data$AllenScore < 4,]
uni_miRs <- unique(tar_data$Query)

miR_mat <- matrix(0, nrow = dim(tpm_data)[1], ncol = length(uni_miRs))

# Populate the miR matrix
for (i in 1:length(uni_miRs)) {
  temp <- unique(tar_data$ids[tar_data$Query %in% uni_miRs[i]])
  genes <- sapply(temp, function(x) unlist(strsplit(x, "\\."))[1])
  miR_mat[match(genes, tpm_data$ID), i] <- 1
}
miR_mat <- data.frame(miR_mat)
rownames(miR_mat) <- tpm_data$ID
colnames(miR_mat) <- uni_miRs
miR_mat$ID <- rownames(miR_mat)

# Reshape miR matrix
miR_mat <- melt(miR_mat, id.vars = c("ID"))
miR_mat$variable <- factor(miR_mat$variable, levels = c("miR161.1", "miR161.2", "miR400", "TAS1a_3'D9(-)", 
                                                        "TAS1b_3'D4(-)", "TAS1c_3'D6(-)", "TAS1c_3'D10(-)", 
                                                        "TAS2_3'D6(-)", "TAS2_3'D9(-)", "TAS2_3'D11(-)", 
                                                        "TAS2_3'D12(-)"))
miR_mat$value <- as.numeric(miR_mat$value)
miR_mat$ID <- factor(as.character(miR_mat$ID))

# Aggregate miR data
miR_tar <- aggregate(value ~ ID, data = miR_mat, FUN = sum)
miR_tar <- transform(miR_tar, 
                      Range_Label = case_when(
                      value == 0 ~ "0",
                      value == 1 ~ "1",
                      value %in% c(2, 3, 4) ~ "2-4",
                      value %in% c(5, 6, 7) ~ "5-7",
                      TRUE ~ "Other"  
                    ))
miR_tar$Count <- 1
miR_tar$Count <- factor(miR_tar$Count)

# Add number of miRNA targets 

p = p + new_scale_fill() +
  geom_fruit(data = miR_tar, 
             geom = geom_tile,
             mapping = aes(y = ID, x = Count, fill = Range_Label),
             color = "white", offset = 0.014, width = 0.2) + 
  scale_fill_manual(name = "Trigger number",
                    values = c("0" = "white", "1" = "#F4CFDD", "2-4" = "#E799B6", "5-7" = "#BE4F81"), 
                    guide = guide_legend(keywidth = 2,keyheight = 2,order = 4))

p = p + new_scale_fill() +
  geom_fruit(data = miR_tar, 
             geom = geom_tile,
             mapping = aes(y = ID, x = Count),
             colour = "black", fill = "black", offset = 0.014,width = 0.02)

# RING4: Add siRNA abundance information
tpm_data <- read.xlsx("TPM_21_PPR_all_accs.xlsx", sheet = 2, sep = "\t")
rownames(tpm_data) <- tpm_data$ID
tpm_data$ID <- NULL

sRNA_temp <- tpm_data
sRNA_temp$Average <- NULL
sRNA_temp[] <- lapply(sRNA_temp, as.numeric)
sRNA_temp[is.na(sRNA_temp)] <- 0
sRNA_temp <- log2(sRNA_temp + 1)

sRNA_temp$ID <- rownames(sRNA_temp)
sRNA_temp <- melt(sRNA_temp, id.vars = "ID", variable.name = "Accession", value.name = "Abundance")
sRNA_temp$Abundance[is.na(sRNA_temp$Abundance)] <- 0

# Visualize the siRNA abundance to the plot
p = p + new_scale_fill() +
  geom_fruit(data = sRNA_temp, geom = geom_tile,
             mapping = aes(y = ID, x = Accession, alpha = Abundance, fill = Accession),
             color = "white", offset = 0.012, size = 0.2) +
  scale_alpha_continuous(range = c(0, 1),
                         guide = guide_legend(keywidth = 2, keyheight = 2, order = 4)) +
  scale_fill_manual(values=c("An-1" = "#0000FF", "Col-0" = "#FFA500", "Ct-1" = "#FF0000",
                             "Cvi-1" = "#800000", "Eri-1" = "#000080", "Kyo-1" = "#008000",
                             "Ler-0" = "#800080", "Sha" = "#696969"),
                    guide = guide_legend(keywidth=2, keyheight = 2, order = 4)) +
  scale_alpha_continuous(name = "TPM (log2)",
                         range = c(0, 1),
                         guide = guide_legend(keywidth = 2, keyheight = 2, order = 5,
                                            legend.text = element_text(size = 20))) +
  theme(legend.position = c(0.95, 0.5),
        legend.key.size = unit(0.5, 'cm'),
        legend.key.height = unit(0.5, 'cm'),
        legend.key.width = unit(0.5, 'cm'),
        legend.background = element_rect(fill = NA),
        legend.title = element_text(size = 20, face = "plain"),
        legend.text = element_text(size = 20, face = "plain"),
        legend.spacing.y = unit(0.5, "cm"),
        legend.spacing.x = unit(0.5, "cm")) +
  theme(plot.margin = unit(c(5, 5, 5, 5), "mm"))

# Save the phylogenetic tree figure
ggsave(plot = p, "ath_475ppr.pep.fa.phy.contree.pdf", width = 100, height = 100, units = "cm", limitsize = FALSE)

```


